README.md
# SftpWrapper

The wrapper for sftp CLI.

_**NOTE: This is an experimental wrapper to download faster than [net-sftp](https://github.com/net-ssh/net-sftp).**_

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'sftp_wrapper'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install sftp_wrapper

## Usage

```ruby
require 'sftp_wrapper'

SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'pass').download('/upload/file', './file')
SftpWrapper::Curl.new('localhost', 2222, 'test', 'pass', curl_args: ['--silent', '--insecure']).download('/upload/file', './file')
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/koshigoe/sftp_wrapper.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
Changelog
====

[v0.2.2](https://github.com/koshigoe/sftp_wrapper/releases/tag/v0.2.2)
----

### Enhancements

- [#3](https://github.com/koshigoe/sftp_wrapper/pull/3) handle `curl` exit status 3.


[v0.2.1](https://github.com/koshigoe/sftp_wrapper/releases/tag/v0.2.1)
----

### Enhancements

- [#2](https://github.com/koshigoe/sftp_wrapper/pull/2) handle more `curl` errors.


[v0.2.0](https://github.com/koshigoe/sftp_wrapper/releases/tag/v0.2.0)
----

### Breaking Change

- [#1](https://github.com/koshigoe/sftp_wrapper/pull/1) Change method signature of `SftpWrapper::Curl#initialize`.


[v0.1.0](https://github.com/koshigoe/sftp_wrapper/releases/tag/v0.1.0)
----

The first release.

- Implement wrapper for `sftp` (OpenSSH)
- Implement wrapper for `curl`


test/test_helper.rb
# frozen_string_literal: true

require 'simplecov'
SimpleCov.start

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)
require 'sftp_wrapper'

require 'minitest/autorun'
require 'tempfile'
require 'net/sftp'

CURL_COMMAND_PATH = ENV['CURL_PREFIX'] ? File.join(ENV['CURL_PREFIX'], 'curl') : 'curl'


test/sftp_wrapper/curl_test.rb
# frozen_string_literal: true

require 'test_helper'

class SftpWrapperCurlTest < Minitest::Test
  describe '#download' do
    describe 'SftpWrapper::Errors::ConnectionError' do
      describe 'wrong port' do
        it 'raise exception' do
          wrapper = SftpWrapper::Curl.new(
            'localhost',
            2223,
            'test',
            'pass',
            curl_path: CURL_COMMAND_PATH,
            curl_args: %w[-s --insecure]
          )
          Tempfile.create('temp') do |temp|
            temp.close
            e = assert_raises(SftpWrapper::Errors::ConnectionError) do
              wrapper.download('/upload/file', temp.path)
            end
            assert_match(/\Aexit status \d+: .*/m, e.message)
          end
        end
      end

      describe 'wrong url' do
        it 'raise exception' do
          wrapper = SftpWrapper::Curl.new(
            'localhost',
            0,
            'test',
            'pass',
            curl_path: CURL_COMMAND_PATH,
            curl_args: %w[-s --insecure]
          )
          Tempfile.create('temp') do |temp|
            temp.close
            e = assert_raises(SftpWrapper::Errors::ConnectionError) do
              wrapper.download('/upload/file', temp.path)
            end
            assert_match(/\Aexit status \d+: .*/m, e.message)
          end
        end
      end

      describe 'wrong host' do
        it 'raise exception' do
          wrapper = SftpWrapper::Curl.new(
            '1234.5.6.7',
            2222,
            'test',
            'pass',
            curl_path: CURL_COMMAND_PATH,
            curl_args: %w[-s --insecure]
          )
          Tempfile.create('temp') do |temp|
            temp.close
            e = assert_raises(SftpWrapper::Errors::ConnectionError) do
              wrapper.download('/upload/file', temp.path)
            end
            assert_match(/\Aexit status \d+: .*/m, e.message)
          end
        end
      end

      describe 'wrong proxy host' do
        it 'raise exception' do
          wrapper = SftpWrapper::Curl.new(
            'localhost',
            2222,
            'test',
            'pass',
            curl_path: CURL_COMMAND_PATH,
            curl_args: %w[-s --insecure --proxy-insecure --proxy 1234.5.6.7]
          )
          Tempfile.create('temp') do |temp|
            temp.close
            e = assert_raises(SftpWrapper::Errors::ConnectionError) do
              wrapper.download('/upload/file', temp.path)
            end
            assert_match(/\Aexit status \d+: .*/m, e.message)
          end
        end
      end
    end

    describe 'SftpWrapper::Errors::AuthenticationFailure' do
      it 'raise exception' do
        wrapper = SftpWrapper::Curl.new(
          'localhost',
          2222,
          'test',
          'passs',
          curl_path: CURL_COMMAND_PATH,
          curl_args: %w[-s --insecure]
        )
        Tempfile.create('temp') do |temp|
          temp.close
          e = assert_raises(SftpWrapper::Errors::AuthenticationFailure) do
            wrapper.download('/upload/file', temp.path)
          end
          assert_match(/\Aexit status \d+: .*/m, e.message)
        end
      end
    end

    describe 'SftpWrapper::Errors::ResourceNotExist' do
      it 'raise exception' do
        wrapper = SftpWrapper::Curl.new(
          'localhost',
          2222,
          'test',
          'pass',
          curl_path: CURL_COMMAND_PATH,
          curl_args: %w[-s --insecure]
        )
        Tempfile.create('temp') do |temp|
          temp.close
          e = assert_raises(SftpWrapper::Errors::ResourceNotExist) do
            wrapper.download('/notfound', temp.path)
          end
          assert_match(/\Aexit status \d+: .*/m, e.message)
        end
      end
    end

    describe 'SftpWrapper::Errors::CommandError' do
      it 'raise exception' do
        wrapper = SftpWrapper::Curl.new(
          'localhost',
          2222,
          'test',
          'pass',
          curl_path: CURL_COMMAND_PATH,
          curl_args: %w[-s --insecureeeeee]
        )
        Tempfile.create('temp') do |temp|
          temp.close
          e = assert_raises(SftpWrapper::Errors::CommandError) do
            wrapper.download('/', temp.path)
          end
          assert_match(/\Aexit status \d+: .*/m, e.message)
        end
      end
    end

    describe 'Success' do
      it 'download file' do
        ssh_options = {
          auth_methods: %w[password],
          config: false,
          non_interactive: true,
          user_known_hosts_file: IO::NULL,
          verify_host_key: :never,
          port: 2222,
          password: 'pass',
        }
        Net::SFTP.start('localhost', 'test', ssh_options) do |session|
          session.open!('/upload/file', 'w') do |res|
            raise unless res.ok?

            session.write!(res[:handle], 0, 'downloaded')
          end
        end

        wrapper = SftpWrapper::Curl.new(
          'localhost',
          2222,
          'test',
          'pass',
          curl_path: CURL_COMMAND_PATH,
          curl_args: %w[-s --insecure]
        )
        Tempfile.create('temp') do |temp|
          temp.close
          wrapper.download('/upload/file', temp.path)
          assert_equal File.read(temp.path), 'downloaded'
        end
      end
    end
  end

  describe '#upload' do
    describe 'SftpWrapper::Errors::ConnectionError' do
      it 'raise exception' do
        wrapper = SftpWrapper::Curl.new(
          'localhost',
          2223,
          'test',
          'pass',
          curl_path: CURL_COMMAND_PATH,
          curl_args: %w[-s --insecure]
        )
        Tempfile.create('temp') do |temp|
          temp.close
          e = assert_raises(SftpWrapper::Errors::ConnectionError) do
            wrapper.upload(temp.path, '/upload/file')
          end
          assert_match(/\Aexit status \d+: .*/m, e.message)
        end
      end
    end

    describe 'SftpWrapper::Errors::AuthenticationFailure' do
      it 'raise exception' do
        wrapper = SftpWrapper::Curl.new(
          'localhost',
          2222,
          'test',
          'passs',
          curl_path: CURL_COMMAND_PATH,
          curl_args: %w[-s --insecure]
        )
        Tempfile.create('temp') do |temp|
          temp.close
          e = assert_raises(SftpWrapper::Errors::AuthenticationFailure) do
            wrapper.upload(temp.path, '/upload/file')
          end
          assert_match(/\Aexit status \d+: .*/m, e.message)
        end
      end
    end

    describe 'SftpWrapper::Errors::CommandError' do
      it 'raise exception' do
        wrapper = SftpWrapper::Curl.new(
          'localhost',
          2222,
          'test',
          'pass',
          curl_path: CURL_COMMAND_PATH,
          curl_args: %w[-s --insecure]
        )

        e = assert_raises(SftpWrapper::Errors::CommandError) do
          wrapper.upload('notfound', '/upload/file')
        end
        assert_match(/\Aexit status \d+: .*/m, e.message)
      end
    end

    describe 'Success' do
      it 'upload file' do
        ssh_options = {
          auth_methods: %w[password],
          config: false,
          non_interactive: true,
          user_known_hosts_file: IO::NULL,
          verify_host_key: :never,
          port: 2222,
          password: 'pass',
        }

        Net::SFTP.start('localhost', 'test', ssh_options) do |session|
          begin
            session.remove!('/upload/file')
          rescue Net::SFTP::StatusException
            nil
          end
        end

        wrapper = SftpWrapper::Curl.new(
          'localhost',
          2222,
          'test',
          'pass',
          curl_path: CURL_COMMAND_PATH,
          curl_args: %w[-s --insecure]
        )
        Tempfile.create('temp') do |temp|
          temp.write('uploaded')
          temp.close

          wrapper.upload(temp.path, '/upload/file')
        end

        Net::SFTP.start('localhost', 'test', ssh_options) do |session|
          assert_equal session.download!('/upload/file', nil), 'uploaded'
        end
      end
    end
  end
end


test/sftp_wrapper/open_ssh_test.rb
# frozen_string_literal: true

require 'test_helper'

class SftpWrapperOpenSSHTest < Minitest::Test
  describe '#download' do
    describe 'SftpWrapper::Errors::ConnectionError' do
      it 'raise exception' do
        wrapper = SftpWrapper::OpenSSH.new('localhost', 2223, 'test', 'pass')
        Tempfile.create('temp') do |temp|
          temp.close
          assert_raises(SftpWrapper::Errors::ConnectionError) do
            wrapper.download('/upload/file', temp.path)
          end
        end
      end
    end

    describe 'SftpWrapper::Errors::AuthenticationFailure' do
      it 'raise exception' do
        wrapper = SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'passs')
        Tempfile.create('temp') do |temp|
          temp.close
          assert_raises(SftpWrapper::Errors::AuthenticationFailure) do
            wrapper.download('/upload/file', temp.path)
          end
        end
      end
    end

    describe 'SftpWrapper::Errors::TimeoutError' do
      it 'raise exception' do
        wrapper = SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'pass', command_timeout: 0)
        Tempfile.create('temp') do |temp|
          temp.close
          assert_raises(SftpWrapper::Errors::TimeoutError) do
            wrapper.download('/upload/file', temp.path)
          end
        end
      end
    end

    describe 'SftpWrapper::Errors::CommandError' do
      it 'raise exception' do
        wrapper = SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'pass')
        Tempfile.create('temp') do |temp|
          temp.close
          assert_raises(SftpWrapper::Errors::CommandError) do
            wrapper.download('/notfound', temp.path)
          end
        end
      end
    end

    describe 'Success' do
      it 'download file' do
        ssh_options = {
          auth_methods: %w[password],
          config: false,
          non_interactive: true,
          user_known_hosts_file: IO::NULL,
          verify_host_key: :never,
          port: 2222,
          password: 'pass',
        }
        Net::SFTP.start('localhost', 'test', ssh_options) do |session|
          session.open!('/upload/file', 'w') do |res|
            raise unless res.ok?

            session.write!(res[:handle], 0, 'downloaded')
          end
        end

        wrapper = SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'pass')
        Tempfile.create('temp') do |temp|
          temp.close
          wrapper.download('/upload/file', temp.path)
          assert_equal File.read(temp.path), 'downloaded'
        end
      end
    end
  end

  describe '#upload' do
    describe 'SftpWrapper::Errors::ConnectionError' do
      it 'raise exception' do
        wrapper = SftpWrapper::OpenSSH.new('localhost', 2223, 'test', 'pass')
        Tempfile.create('temp') do |temp|
          temp.close
          assert_raises(SftpWrapper::Errors::ConnectionError) do
            wrapper.upload(temp.path, '/upload/file')
          end
        end
      end
    end

    describe 'SftpWrapper::Errors::AuthenticationFailure' do
      it 'raise exception' do
        wrapper = SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'passs')
        Tempfile.create('temp') do |temp|
          temp.close
          assert_raises(SftpWrapper::Errors::AuthenticationFailure) do
            wrapper.upload(temp.path, '/upload/file')
          end
        end
      end
    end

    describe 'SftpWrapper::Errors::TimeoutError' do
      it 'raise exception' do
        wrapper = SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'pass', command_timeout: 0)
        Tempfile.create('temp') do |temp|
          temp.close
          assert_raises(SftpWrapper::Errors::TimeoutError) do
            wrapper.upload(temp.path, '/upload/file')
          end
        end
      end
    end

    describe 'SftpWrapper::Errors::CommandError' do
      it 'raise exception' do
        wrapper = SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'pass')

        assert_raises(SftpWrapper::Errors::CommandError) do
          wrapper.upload('notfound', '/upload/file')
        end
      end
    end

    describe 'Success' do
      it 'upload file' do
        ssh_options = {
          auth_methods: %w[password],
          config: false,
          non_interactive: true,
          user_known_hosts_file: IO::NULL,
          verify_host_key: :never,
          port: 2222,
          password: 'pass',
        }

        Net::SFTP.start('localhost', 'test', ssh_options) do |session|
          begin
            session.remove!('/upload/file')
          rescue Net::SFTP::StatusException
            nil
          end
        end

        wrapper = SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'pass')
        Tempfile.create('temp') do |temp|
          temp.write('uploaded')
          temp.close

          wrapper.upload(temp.path, '/upload/file')
        end

        Net::SFTP.start('localhost', 'test', ssh_options) do |session|
          assert_equal session.download!('/upload/file', nil), 'uploaded'
        end
      end
    end
  end
end


lib/sftp_wrapper/curl.rb
# frozen_string_literal: true

require 'uri'
require 'erb'
require 'open3'

module SftpWrapper
  # The wrapper for curl.
  class Curl
    attr_reader :host, :port, :username, :password, :curl_path, :curl_args

    # lookup table of curl errors.
    CURL_ERRORS = {
      # URL malformed. The syntax was not correct.
      3 => SftpWrapper::Errors::ConnectionError,
      # Couldn't resolve proxy. The given proxy host could not be resolved.
      5 => SftpWrapper::Errors::ConnectionError,
      # Couldn't resolve host. The given remote host was not resolved.
      6 => SftpWrapper::Errors::ConnectionError,
      # Failed to connect to host.
      7 => SftpWrapper::Errors::ConnectionError,
      # The user name, password, or similar was not accepted and curl failed to log in.
      67 => SftpWrapper::Errors::AuthenticationFailure,
      # The resource referenced in the URL does not exist.
      78 => SftpWrapper::Errors::ResourceNotExist,
    }.freeze

    # Initialize SFTP wrapper.
    #
    # @param host [String] host address of SFTP server
    # @param port [Integer] port number of SFTP server
    # @param username [String] user name of SFTP server
    # @param password [String] password of SFTP server
    # @param options [Hash] curl options.
    # @option options [String] :curl_path path of `curl` command.
    # @option options [Array<String>] :curl_args command line arguments of `curl`.
    #
    def initialize(host, port, username, password, options = {})
      @host = host
      @port = port
      @username = username
      @password = password
      @curl_path = options[:curl_path] || 'curl'
      @curl_args = options[:curl_args] || []
    end

    # Get remote file.
    #
    # @param source [String] source file path
    # @param destination [String] destination path
    # @raise [SftpWrapper::Errors::ConnectionError]
    # @raise [SftpWrapper::Errors::AuthenticationFailure]
    # @raise [SftpWrapper::Errors::CommandError]
    #
    def download(source, destination)
      userinfo = [username, password].map(&ERB::Util.method(:url_encode)).join(':')
      uri = URI::Generic.build(scheme: 'sftp', userinfo: userinfo, host: host, port: port, path: source)
      cmd = %W[#{curl_path} -o #{destination}] + curl_args + [uri.to_s]

      execute(*cmd)
    end

    # Put local file.
    #
    # @param source [String] source file path
    # @param destination [String] destination path
    # @raise [SftpWrapper::Errors::ConnectionError]
    # @raise [SftpWrapper::Errors::AuthenticationFailure]
    # @raise [SftpWrapper::Errors::CommandError]
    #
    def upload(source, destination)
      userinfo = [username, password].map(&ERB::Util.method(:url_encode)).join(':')
      uri = URI::Generic.build(scheme: 'sftp', userinfo: userinfo, host: host, port: port, path: destination)
      cmd = %W[#{curl_path} -T #{source}] + curl_args + [uri.to_s]

      execute(*cmd)
    end

    private

    def execute(*cmd)
      _, stderr, status = Open3.capture3(*cmd)

      return if status.success?

      exception_class = CURL_ERRORS[status.exitstatus] || SftpWrapper::Errors::CommandError

      raise exception_class, "exit status #{status.exitstatus}: #{stderr}"
    end
  end
end


lib/sftp_wrapper/open_ssh.rb
# frozen_string_literal: true

require 'pty'
require 'expect'
require 'shellwords'

module SftpWrapper
  # The wrapper for OpenSSH's sftp command.
  class OpenSSH
    attr_reader :host,
                :port,
                :username,
                :password,
                :ssh_options,
                :ssh_config,
                :open_timeout,
                :read_timeout,
                :command_timeout,
                :debug

    # Default value of open timeout.
    DEFAULT_OPEN_TIMEOUT = 10
    # Default value of read timeout.
    DEFAULT_READ_TIMEOUT = 10
    # Default value of command timeout.
    DEFAULT_COMMAND_TIMEOUT = 9_999_999
    # Default value of ssh options
    DEFAULT_SSH_OPTIONS = {
      PasswordAuthentication: 'yes',
      PreferredAuthentications: 'password',
      PubkeyAuthentication: 'no',
      NumberOfPasswordPrompts: 1,
      StrictHostKeyChecking: 'no',
      UserKnownHostsFile: '/dev/null',
      TCPKeepAlive: 'yes',
      ServerAliveInterval: 60,
      ServerAliveCountMax: 3,
    }.freeze
    # Default value of ssh config file path.
    DEFAULT_SSH_CONFIG = '/dev/null'
    # pattern of password prompt.
    PASSWORD_PROMPT_RE = / password: /.freeze
    # sftp prompt.
    SFTP_PROMPT = 'sftp> '
    # pattern of sftp prompt.
    SFTP_PROMPT_RE = /^#{SFTP_PROMPT}/.freeze

    # Initialize SFTP wrapper.
    #
    # @param host [String] host address of SFTP server
    # @param port [Integer] port number of SFTP server
    # @param username [String] user name of SFTP server
    # @param password [String] password of SFTP server
    # @param ssh_options [Hash] SSH options (set as -o options)
    # @param ssh_config [String] path of SSH config file (set as -F option unless nil)
    # @param open_timeout [Integer, Float]
    # @param read_timeout [Integer, Float]
    # @param command_timeout [Integer, Float]
    # @param debug [Boolean]
    #
    # rubocop:disable Metrics/ParameterLists
    def initialize(host, port, username, password,
                   ssh_options: {},
                   ssh_config: DEFAULT_SSH_CONFIG,
                   open_timeout: DEFAULT_OPEN_TIMEOUT,
                   read_timeout: DEFAULT_READ_TIMEOUT,
                   command_timeout: DEFAULT_COMMAND_TIMEOUT,
                   debug: false)
      @host = host
      @port = port
      @username = username
      @password = password
      @ssh_options = DEFAULT_SSH_OPTIONS.merge(ssh_options)
      @ssh_config = ssh_config
      @open_timeout = open_timeout
      @read_timeout = read_timeout
      @command_timeout = command_timeout
      @debug = debug
    end
    # rubocop:enable Metrics/ParameterLists

    # Get remote file.
    #
    # @param source [String] source file path
    # @param destination [String] destination path
    # @raise [SftpWrapper::Errors::ConnectionError]
    # @raise [SftpWrapper::Errors::AuthenticationFailure]
    # @raise [SftpWrapper::Errors::TimeoutError]
    # @raise [SftpWrapper::Errors::CommandError]
    #
    def download(source, destination)
      execute(%W[get #{source} #{destination}].shelljoin)
    end

    # Put local file.
    #
    # @param source [String] source file path
    # @param destination [String] destination path
    # @raise [SftpWrapper::Errors::ConnectionError]
    # @raise [SftpWrapper::Errors::AuthenticationFailure]
    # @raise [SftpWrapper::Errors::TimeoutError]
    # @raise [SftpWrapper::Errors::CommandError]
    #
    def upload(source, destination)
      execute(%W[put #{source} #{destination}].shelljoin)
    end

    private

    # Execute sftp command.
    #
    # @param command [String] SFTP command
    # @raise [SftpWrapper::Errors::ConnectionError]
    # @raise [SftpWrapper::Errors::AuthenticationFailure]
    # @raise [SftpWrapper::Errors::TimeoutError]
    # @raise [SftpWrapper::Errors::CommandError]
    #
    def execute(command) # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
      cli = %w[sftp] + build_cli_args + [host]
      PTY.getpty(cli.shelljoin) do |r, w, _pid|
        w.sync = true

        unless safe_expect(r, PASSWORD_PROMPT_RE, open_timeout)
          raise SftpWrapper::Errors::ConnectionError, 'connection error.'
        end

        w.puts(password)

        unless safe_expect(r, SFTP_PROMPT_RE, read_timeout)
          raise SftpWrapper::Errors::AuthenticationFailure, 'authentication failure'
        end

        w.puts(command)

        res = safe_expect(r, SFTP_PROMPT_RE, command_timeout)
        raise SftpWrapper::Errors::TimeoutError, 'command timed out' unless res

        skip = command.bytesize + 2
        error_message = res[0].byteslice(skip..-1).chomp.sub(/#{SFTP_PROMPT}\z/, '').chomp
        raise SftpWrapper::Errors::CommandError, error_message unless error_message.empty?

        w.puts('quit')
      end
    end

    def build_cli_args
      args = []
      args << '-q'
      args += %W[-F #{ssh_config}] if ssh_config
      args += %W[-P #{port} -o User=#{username}]
      args += ssh_options.map { |k, v| %W[-o #{k}=#{v}] }.flatten
      args
    end

    # @see https://github.com/ruby/ruby/blob/13b692200dba1056fa9033f2c64c43453f6d6a98/ext/pty/pty.c#L716-L723
    def safe_expect(io, pattern, timeout)
      expect_verbose = $expect_verbose # rubocop:disable Style/GlobalVars
      $expect_verbose = debug # rubocop:disable Style/GlobalVars
      io.expect(pattern, timeout)
    rescue Errno::EIO
      nil
    ensure
      $expect_verbose = expect_verbose # rubocop:disable Style/GlobalVars
    end
  end
end


benchmark/benchmark_upload.rb
# frozen_string_literal: true

require 'benchmark'
require 'net/sftp'
require 'sftp_wrapper'
require 'tempfile'

data = ' ' * (1024 * 1024 * 100)

Tempfile.create('temp') do |temp|
  temp.write(data)
  temp.close

  Benchmark.bm(16) do |x|
    x.report('OpenSSH sftp CLI') do
      wrapper = SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'pass')
      wrapper.upload(temp.path, '/upload/benchmark-download')
    end

    x.report('curl CLI') do
      wrapper = SftpWrapper::Curl.new('localhost', 2222, 'test', 'pass')
      wrapper.upload(temp.path, '/upload/benchmark-download')
    end

    x.report('net-sftp') do
      ssh_options = {
        auth_methods: %w[password],
        config: false,
        non_interactive: true,
        user_known_hosts_file: IO::NULL,
        verify_host_key: :never,
        port: 2222,
        password: 'pass',
      }
      Net::SFTP.start('localhost', 'test', ssh_options) do |session|
        session.upload!(temp.path, '/upload/benchmark-upload')
      end
    end
  end
end

__END__

$ ruby -v
ruby 2.6.4p104 (2019-08-28 revision 67798) [x86_64-linux]
$ bundle exec ruby benchmark_upload.rb
                       user     system      total        real
OpenSSH sftp CLI   0.004058   0.000321   0.004379 (  1.749079)
curl CLI           0.007233   0.003560   0.676544 (  2.184746)
net-sftp           0.852571   0.099383   0.951954 (  2.413606)


benchmark/benchmark_download.rb
# frozen_string_literal: true

require 'benchmark'
require 'net/sftp'
require 'sftp_wrapper'
require 'tempfile'

data = ' ' * (1024 * 1024 * 100)

ssh_options = {
  auth_methods: %w[password],
  config: false,
  non_interactive: true,
  user_known_hosts_file: IO::NULL,
  verify_host_key: :never,
  port: 2222,
  password: 'pass',
}

Tempfile.create('temp') do |temp|
  temp.write(data)
  temp.close

  Net::SFTP.start('localhost', 'test', ssh_options) do |session|
    session.upload!(temp.path, '/upload/benchmark-download')
  end
end

Benchmark.bm(16) do |x|
  x.report('OpenSSH sftp CLI') do
    Tempfile.create('temp') do |temp|
      temp.close

      wrapper = SftpWrapper::OpenSSH.new('localhost', 2222, 'test', 'pass')
      wrapper.download('/upload/benchmark-download', 'downloaded')
    end
  end

  x.report('curl CLI') do
    Tempfile.create('temp') do |temp|
      temp.close

      wrapper = SftpWrapper::Curl.new('localhost', 2222, 'test', 'pass')
      wrapper.download('/upload/benchmark-download', 'downloaded')
    end
  end

  x.report('net-sftp') do
    Tempfile.create('temp') do |temp|
      temp.close

      Net::SFTP.start('localhost', 'test', ssh_options) do |session|
        session.download!('/upload/benchmark-download', temp.path)
      end
    end
  end
end

__END__

$ ruby -v
ruby 2.6.4p104 (2019-08-28 revision 67798) [x86_64-linux]
$ bundle exec ruby benchmark_download.rb
                       user     system      total        real
OpenSSH sftp CLI   0.000382   0.003365   0.003747 (  3.044737)
curl CLI           0.015933   0.000000   2.115820 (  3.375170)
net-sftp           6.210004   0.744365   6.954369 ( 10.792968)
